sympy/ntheory/residue_ntheory.py

```diff
@@ -427,90 +429,49 @@ def _sqrt_mod_prime_power(a, p, k):
-            # then r + 2**(nx - 1) is a root mod 2**(nx+1)
-            n = 3
-            res = []
-            for r in rv:
-                nx = n
-                while nx < k:
-                    r1 = (r**2 - a) >> nx
-                    if r1 % 2:
-                        r = r + (1 << (nx - 1))
-                    #assert (r**2 - a)% (1 << (nx + 1)) == 0
-                    nx += 1
-                if r not in res:
-                    res.append(r)
-                x = r + (1 << (k - 1))
-                #assert (x**2 - a) % pk == 0
-                if x < (1 << nx) and x not in res:
-                    if (x**2 - a) % pk == 0:
-                        res.append(x)
-            return res
-        rv = _sqrt_mod_prime_power(a, p, 1)
-        if not rv:
+        if a % 8 != 1:
             return None
-        r = rv[0]
-        fr = r**2 - a
-        # hensel lifting with Newton iteration, see Ref.[3] chapter 9
+        # Trivial
+        if k <= 3:
+            return [ZZ(i) for i in range(1, pk, 2)]
+        r = 1
+        # r is one of the solutions to x**2 - a = 0 (mod 2**3).
+        # Hensel lift them to solutions of x**2 - a = 0 (mod 2**k)
+        # if r**2 - a = 0 mod 2**nx but not mod 2**(nx+1)
+        # then r + 2**(nx - 1) is a root mod 2**(nx+1)
+        for nx in range(3, k):
+            if ((r**2 - a) >> nx) % 2:
+                r += 1 << (nx - 1)
+        # r is a solution of x**2 - a = 0 (mod 2**k), and
+        # there exist other solutions -r, r+h, -(r+h), and these are all solutions.
+        h = 1 << (k - 1)
+        return sorted([r, pk - r, (r + h) % pk, -(r + h) % pk])
+
+    # If the Legendre symbol (a/p) is not 1, no solution exists.
+    if jacobi_symbol(a, p) != 1:
+        return None
+    if p % 4 == 3:
+        res = pow(a, (p + 1) // 4, p)
+    elif p % 8 == 5:
+        res = pow(a, (p + 3) // 8, p)
+        if pow(res, 2, p) != a % p:
+            res = res * pow(2, (p - 1) // 4, p) % p
+    else:
+        res = _sqrt_mod_tonelli_shanks(a, p)
+    if k > 1:
+        # Hensel lifting with Newton iteration, see Ref.[3] chapter 9
         # with f(x) = x**2 - a; one has f'(a) != 0 (mod p) for p != 2
-        n = 1
         px = p
-        while 1:
-            n1 = n
-            n1 *= 2
-            if n1 > k:
-                break
-            n = n1
+        for _ in range(k.bit_length() - 1):
             px = px**2
-            frinv = igcdex(2*r, px)[0]
-            r = (r - fr*frinv) % px
-            fr = r**2 - a
-        if n < k:
-            px = p**k
-            frinv = igcdex(2*r, px)[0]
-            r = (r - fr*frinv) % px
-        return [r, px - r]
+            frinv = igcdex(2*res, px)[0]
+            res = (res - (res**2 - a)*frinv) % px
+        if k & (k - 1): # If k is not a power of 2
+            frinv = igcdex(2*res, pk)[0]
+            res = (res - (res**2 - a)*frinv) % pk
+    return sorted([ZZ(res), ZZ(pk - res)])
```

Comment:
At 450 we are not returning ZZ results -- are we returning different types from here?

Reply:
The current code also mixes `int` and `ZZ`, which is not a problem, but it would be safer to align them.

Reply:
I would have a question why it had been reviewed with a mix of `ZZ` and `int` at the first place.
But for `ntheory` module, I just expect everything to be `int`

I think that dropping `ZZ` constructor would be better. 
Even if we want to assure some compatiblity with gmpy data types, they should be compatible with the argument of object algebra.

Reply:
I would have thought that the ntheory module should definitely use gmpy2 if it is installed. It is much faster for large integers. Separately from polys we have MPZ:
https://github.com/sympy/sympy/blob/6f2e2192d07b7bdf239b8b48b9b3afadc5cb42e0/sympy/external/gmpy.py#L28-L29
Also some functions provided by gmpy2 should be used in ntheory when it is installed.

Reply:
There is none or little gain with efficiency when using explicit constructor of `ZZ`, 
and if users want to get `gmpy` result, then we only have to guarantee that gmpy results are get when gmpy objects are used (not when python int is used), and that is guaranteed just by dropping the constructor
and justified by the [object algebra](https://www.cs.utexas.edu/~wcook/Drafts/2012/ecoop2012.pdf)

Reply:
Yes, I suppose just converting with `ZZ` on output is not that useful.

Reply:
I think that it is better to switch the functions, than the data types
https://gmpy2.readthedocs.io/en/latest/mpz.html#mpz-functions

Reply:
Unified return type to int